fili-core/src/main/java/com/yahoo/bard/webservice/table/MetricUnionCompositeTable.java

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.data.time.ZonedTimeGrain;
+import com.yahoo.bard.webservice.table.availability.MetricUnionAvailability;
+import com.yahoo.bard.webservice.table.resolver.GranularityComparator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An implementation of <tt>BasePhysicalTable</tt> backed by metric union availability.
+ */
+public class MetricUnionCompositeTable extends BasePhysicalTable {
+    /**
+     * Constructor.
+     *
+     * @param name  Name of the physical table as TableName, also used as fact table name
+     * @param columns  The columns for this table
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>s
+     * @param logicalToPhysicalColumnNames  Mappings from logical to physical names
+     */
+    public MetricUnionCompositeTable(
+            @NotNull TableName name,
+            @NotNull Set<Column> columns,
+            @NotNull Set<PhysicalTable> physicalTables,
+            @NotNull Map<String, String> logicalToPhysicalColumnNames
+    ) {
+        super(
+                name,
+                getCoarsestTimeGrain(physicalTables),
+                columns,
+                logicalToPhysicalColumnNames,
+                new MetricUnionAvailability(physicalTables, columns)
+        );
+    }
+
+    /**
+     * Returns the coarsest <tt>ZonedTimeGrain</tt> that satisfies all tables.
+     * <p>
+     * If the set of <tt>PhysicalTables</tt>'s is empty or the coarsest <tt>ZonedTimeGrain</tt> is not
+     * compatible with any of the <tt>PhysicalTables</tt>s, throw <tt>IllegalArgumentException</tt>.
+     *
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>s among which the coarsest <tt>ZonedTimeGrain</tt>
+     * is to be returned.
+     *
+     * @return the coarsest <tt>ZonedTimeGrain</tt> among a set of <tt>PhysicalTables</tt>s
+     */
+    private static ZonedTimeGrain getCoarsestTimeGrain(Set<PhysicalTable> physicalTables) {
+        if (physicalTables.isEmpty()) {
+            throw new IllegalArgumentException("At least 1 physical table needs to be provided");
+        }
```

Comment:
@cdeszaq I like this and really want to do it this way. [A compiler bug](http://stackoverflow.com/questions/39076077/throw-runtimeexception-inside-stream-with-optional-orelsethrow), however, stops me from doing this :( Let's talk about this offline